feat: add curve executor

royvardhan opened the pull request.

tamaralipows commented on foundry/src/executors/CurveExecutor.sol:
```diff
@@ -0,0 +1,62 @@
+    using SafeERC20 for IERC20;
+
+    ICurveRouter public immutable curveRouter;
+    address public immutable ethAddress;
```
We expect this to work on other EVM-compatible chains right? In that case it should probably we native token, not eth?

tamaralipows commented on foundry/test/executors/CurveExecutor.t.sol:
```diff
@@ -0,0 +1,594 @@
+        returns (int128, int128, bool);
+}
+
+// Aave lending pool
```
These comments are not very meaningful imo, this info is already in the interface name

tamaralipows commented on foundry/test/executors/CurveExecutor.t.sol:
```diff
@@ -0,0 +1,594 @@
+// This is the registry that contains the information about the pool
+// The naming convention is different because it is in vyper
+interface MetaRegistry {
+    // Get the number of coins in the pool
+    function get_n_coins(address pool) external view returns (uint256);
+    // Get the indices of the coins in the pool
+    function get_coin_indices(address pool, address from, address to)
```
Same with these method comments - they don't add any additional info. It's a concept in the [book](chrome-extension://efaidnbmnnnibpcajpcglclefindmkaj/https://milkov.tech/assets/psd.pdf) - by the way, under the section `Comment Repeats Code`.

The interface comment is fine

tamaralipows commented on foundry/test/executors/CurveExecutor.t.sol:
```diff
@@ -0,0 +1,594 @@
+        assertEq(params.swapParams[0][3], 3);
+    }
+
+    // The following pools are unique and do not have a factory
```
What does it mean if they don't have a factory? Do we have to execute differently for these pools?

dianacarvalho1 replied:
we might. not having a factory means that the curve devs decided to deploy a pool on its own just because. and it has unique code 😬 it's terrible

tamaralipows commented on foundry/test/executors/CurveExecutor.t.sol:
```diff
@@ -0,0 +1,594 @@
+    }
+
+    // The following pool is from CryptoSwapNG, deployed by factory 0x6A8cbed756804B16E05E741eDaBd5cB544AE21bf
+    // - It is a plain pool
```
What does plain pool vs meta pool mean? For my own knowledge

tamaralipows commented on foundry/test/executors/CurveExecutor.t.sol:
```diff
@@ -0,0 +1,594 @@
+        );
+    }
+
+    // The following pool is from Tricrypto, deployed by factory 0x0c0e5f2fF0ff18a3be9b835635039256dC4B4963
```
Why do we need to know these factory addresses?

royvardhan replied:
Nothing, I took these comments from the indexer.

tamaralipows commented on foundry/test/executors/CurveExecutor.t.sol:
```diff
@@ -0,0 +1,594 @@
+        );
+    }
+
+    // The following pools were taken from the old curve executor tests
```
Not a meaningful comment since future devs have no idea what the old tests are. Better to write what they are actually testing

royvardhan replied:
agreed!

tamaralipows commented on foundry/test/executors/CurveExecutor.t.sol:
```diff
@@ -0,0 +1,594 @@
+        swapParams[0][4] = nCoins;
+    }
+
+    function dealAaveDai() internal {
```
This function is only used in one place AFAIU - is it worth to make it a separate function?

tamaralipows reviewed (commented):
Thanks for this! I have mostly questions for my own understanding 🙏

tamaralipows commented on foundry/src/executors/CurveExecutor.sol:
```diff
@@ -0,0 +1,62 @@
+        ICurveRouter.CurveRouterParams memory params = _decodeData(data);
+        if (params.route[0] != ethAddress) {
+            // slither-disable-next-line unused-return
+            IERC20(params.route[0]).approve(address(curveRouter), amountIn);
```
Missed one thing - we will actually be doing dynamic approvals here to save on gas (i.e. we approve once for MAX if we need to, otherwise we don't approve anything).

Please have a look at how this is done in Balancer!

dianacarvalho1 replied:
you need to pass a bool from encoding to decide if we need to approve or not!

royvardhan replied:
Yeah I forgot, updating tests to supply a bool which indicates if approval is needed or not.

dianacarvalho1 commented on foundry/test/executors/CurveExecutor.t.sol:
```diff
@@ -0,0 +1,594 @@
+        address[11] memory route =
+            _getRoute(WETH_ADDR, USDC_ADDR, TRICRYPTO_POOL);
+
+        // The meta registry does not have information about the pool.
```
why is this relevant? Can I have a bit more context please? 😅  what does the meta registry have to do with anything here?

royvardhan replied:
Meta registry helps us get `n_coins` and `coin_indices` for any registered curve pool. And these vals are needed for a swap. For all the other tests I am using `_getSwapParams` which uses the meta registry, but for this particular pool it is reverting.

dianacarvalho1 commented on foundry/test/executors/CurveExecutor.t.sol:
```diff
@@ -0,0 +1,594 @@
+
+        uint256 amountOut = curveExecutorExposed.swap(amountIn, data);
+
+        assertTrue(amountOut >= 1 ether);
```
can we have a = check instead of a >= please?

dianacarvalho1 commented on foundry/test/executors/CurveExecutor.t.sol:
```diff
@@ -0,0 +1,594 @@
+        uint256 amountOut = curveExecutorExposed.swap(amountIn, data);
+
+        assertTrue(amountOut >= 1 ether);
+        assertEq(IERC20(STETH_ADDR).balanceOf(address(this)), amountOut - 1); // Gets 1 wei less than amountOut
```
why?

dianacarvalho1 commented on foundry/test/executors/CurveExecutor.t.sol:
```diff
@@ -0,0 +1,594 @@
+        );
+    }
+
+    // The following pool is from CryptoSwapNG, deployed by factory 0x6A8cbed756804B16E05E741eDaBd5cB544AE21bf
```
can we have these comments inside of the test please? not outside

dianacarvalho1 reviewed (commented):
Thank you so much @royvardhan ! 🙏🏼  this looks great! I have some comments and one overall question:
How did you figure out which pool type (from the router types) were the pools?

tamaralipows commented on foundry/src/executors/CurveExecutor.sol:
```diff
@@ -0,0 +1,62 @@
+    address public immutable ethAddress;
+
+    constructor(address _curveRouter, address _ethAddress) {
+        if (_curveRouter == address(0) || _ethAddress == address(0)) {
+            revert CurveExecutor__InvalidAddresses();
+        }
```
no? || isn't XOR?

royvardhan replied:
if any of these vals are zero address then we revert. If the first val is zero, it will revert without checking the other val.

tamaralipows commented:
I see you are also having problems with slither, this is how I fixed it in the feature branch: https://github.com/propeller-heads/tycho-execution/pull/117/commits/f57cc213ec7868c886f3d99a66d354439f853d28

royvardhan commented:
> I see you are also having problems with slither, this is how I fixed it in the feature branch: [f57cc21](https://github.com/propeller-heads/tycho-execution/commit/f57cc213ec7868c886f3d99a66d354439f853d28)

thanks! fixed it.

tamaralipows reviewed (commented):
Thanks for this @royvardhan 

I have some minor questions, though my biggest concern is that I don't find the following [test cases](https://github.com/propeller-heads/tycho-protocol-sdk/blob/main/substreams/ethereum-curve/integration_test.tycho.yaml) here:

1. Metapool factory 0xB9fC157394Af804a3578134A6585C0dc9cc990d4 - PlainPool
2. CryptoPool factory 0xF18056Bbd320E96A48e3Fbf8bC061322531aac99 - with ETH
  
Could you please let me know if I've missed them?

tamaralipows commented on foundry/src/executors/CurveExecutor.sol:
```diff
@@ -0,0 +1,68 @@
+        uint256[5][5] swapParams;
+        uint256 amountIn;
+        uint256 minAmountOut;
+        address[5] pools;
```
If this isn't needed right now then I say we don't need to decode it. Instead, we can perhaps just hardcode it as input to the exchange method

royvardhan replied:
yes sure. Thanks for this suggestion, i am removing it

royvardhan commented on foundry/test/executors/CurveExecutor.t.sol:
```diff
@@ -0,0 +1,626 @@
+        );
+    }
+
+    function testSwapAavePool() public {
```
Yes, they were carried over from the previous repo. If we are not indexing them anymore then we should remove them

royvardhan replied:
I am removing them

royvardhan commented:
> Thanks for this @royvardhan
> 
> I have some minor questions, though my biggest concern is that I don't find the following [test cases](https://github.com/propeller-heads/tycho-protocol-sdk/blob/main/substreams/ethereum-curve/integration_test.tycho.yaml) here:
> 
> 1. Metapool factory 0xB9fC157394Af804a3578134A6585C0dc9cc990d4 - PlainPool
> 2. CryptoPool factory 0xF18056Bbd320E96A48e3Fbf8bC061322531aac99 - with ETH
> 
> Could you please let me know if I've missed them?

You have mentioned the factory address, but I get your point, these are the following pools that I haven't added in tests due to significantly low liquidity:

royvardhan commented:
> Thanks for this @royvardhan
> 
> I have some minor questions, though my biggest concern is that I don't find the following [test cases](https://github.com/propeller-heads/tycho-protocol-sdk/blob/main/substreams/ethereum-curve/integration_test.tycho.yaml) here:
> 
> 1. Metapool factory 0xB9fC157394Af804a3578134A6585C0dc9cc990d4 - PlainPool
> 2. CryptoPool factory 0xF18056Bbd320E96A48e3Fbf8bC061322531aac99 - with ETH
> 
> Could you please let me know if I've missed them?

You have mentioned the factory address, but I get your point, these are the following pools that I haven't added in tests due to significantly low liquidity:

1. [wstTAO/wTAO](https://etherscan.io/address/0xf2DCf6336D8250754B4527f57b275b19c8D5CF88) - $5k liquidity
2. [pxETH/mkUS](https://etherscan.io/address/0x71db3764d6841d8b01dc27c0fd4a66a8a34b2be0) - $178 liquidity

royvardhan commented on foundry/test/executors/CurveExecutor.t.sol:
```diff
@@ -0,0 +1,626 @@
+
+        uint256 amountIn = 1 ether;
+        uint256 minAmountOut = 0;
+        address[5] memory pools;
```
resolving notes: had a discussion about this In chat

royvardhan commented on foundry/test/executors/CurveExecutor.t.sol:
```diff
@@ -0,0 +1,626 @@
+        );
+    }
+
+    function _getRoute(address tokenIn, address tokenOut, address pool)
+        internal
+        pure
+        returns (address[11] memory route)
+    {
+        route[0] = tokenIn;
+        route[2] = tokenOut;
+        route[1] = pool;
+    }
```
You can do upto 5 swaps in a single call.

royvardhan replied:
- tokenIn
 - pool
 - tokenOut
 - pool
 - tokenOut
 - pool
 - tokenOut
 - pool
 - tokenOut
 - pool
 - tokenOut

tamaralipows commented:
We still support those pool types despite the low liquidity, don't we?

tamaralipows commented on foundry/test/executors/CurveExecutor.t.sol:
```diff
@@ -252,7 +252,7 @@ contract CurveExecutorTest is Test, Constants {
 
     function testSwapWethXyoPool() public {
-        // The following pool is from CryptoPool, deployed by factory 0xF18056Bbd320E96A48e3Fbf8bC061322531aac99
+        // The following pool is from CryptoPool, deployed by factory 0xF18056Bbd320E96A48e3Fbf8bC061322531aac99 - with ETH
```
I'm a bit confused - is this a `WETH/XYO` pool or an `ETH/XYO` pool? In the sdk integration tests it seems to be an `ETH/XYO` pool but here you are testing with `WETH`. Does it support both? I think this test case should test ETH since you have already added a non-ETH test case, right?

tamaralipows commented on foundry/src/executors/CurveExecutor.sol:
```diff
@@ -0,0 +1,67 @@
+    struct SwapParams {
+        address[11] route;
+        uint256[5][5] swapParams;
+        uint256 minAmountOut;
```
We should remove this from the swap params since we aren't checking min amount out at the executor level.

tamaralipows reviewed (commented):
Thanks - looks like we are almost there! See my two comments please

dianacarvalho1 commented on foundry/src/executors/CurveExecutor.sol:
```diff
@@ -0,0 +1,133 @@
+
+error CurveExecutor__InvalidAddresses();
+
+interface CryptoPool {
```
do you think these should be in a lib? I kinda felt like no. we literally only need them here

tamaralipows replied:
no I think this is fine

dianacarvalho1 commented on foundry/src/executors/CurveExecutor.sol:
```diff
@@ -0,0 +1,133 @@
+error CurveExecutor__InvalidAddresses();
+
+interface CryptoPool {
+    // slither-disable-next-line naming-convention
```
I decided to keep the naming like it was in Vyper (the language curve is written in)

dianacarvalho1 commented on foundry/test/executors/CurveExecutor.t.sol:
```diff
@@ -0,0 +1,351 @@
+        return (coinInIndex, coinOutIndex);
+    }
+
+    function dealAaveDai() internal {
```
ah I see this function is useless now. I will delete it tomorrow!

dianacarvalho1 commented on foundry/test/Constants.sol:
```diff
@@ -30,6 +30,7 @@ contract Constants is Test, BaseConstants {
     address UNPAUSER = makeAddr("unpauser");
 
     // Assets
+    address ETH_ADDR = address(0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE);
```
this worries me 😬 
I've tried to run the tests with ETH being the zero address and it works for all of them except the stETH pool 😬 this one it really needs to be this eeeeeEee address. @tamaralipows Do you think it's ok to just address this in the encoder? Meaning, to see if it's this one stupid pool and if it's ETH. If yes, change the token address to EeeeeE

dianacarvalho1 replied:
I ended up fixing this in the last commit of the encoding PR [here](https://github.com/propeller-heads/tycho-execution/pull/153/commits/7cc0e8f58699cd487ce39c08dd45d4731cf6fb8c)

dianacarvalho1 commented on foundry/src/executors/CurveExecutor.sol:
```diff
@@ -0,0 +1,133 @@
+        returns (uint256)
+    {
+        (
+            address tokenIn,
+            address tokenOut,
+            address pool,
+            uint8 poolType,
+            int128 i,
+            int128 j,
+            bool tokenApprovalNeeded
```
notice that there isn't a receiver 😕 We will never be able to optimize the transfers from a curve pool. it will always have to go through the router

tamaralipows reviewed (approved):
Thank you for this! It looks so clean now!

dianacarvalho1 commented on foundry/src/TychoRouter.sol:
```diff
@@ -1109,13 +1110,35 @@ contract TychoRouter is AccessControl, Dispatcher, EIP712 {
             int256 outputDelta = _getDelta(tokenOut);
+            console2.log("outputDelta:", outputDelta);
+            console2.log("required contribution:", requiredContribution);
             if (outputDelta > 0) {
```
why couldn't we just do `if (outputDelta > 0 || receiver == address(this) )`
Just tried this now quickly and I think it would work 🤔 and it's simpler than crediting here

dianacarvalho1 reviewed (commented):
Thank you @tamaralipows 🙏🏼 

> Should we put a hard cap on the client contribution?

I'm thinking yes. It should never be higher than the minAmountOut.

> Should we tie the client contribution signature to the input amount?

So that people can't reuse the client contribution across different swaps? Initially we thought about tying to the amount in and token in but then decided not to. Can't remember exactly why now..

dianacarvalho1 commented on foundry/test/TychoRouterVault.t.sol:
```diff
@@ -346,6 +346,132 @@ contract TychoRouterUsingVaultTest is TychoRouterTestSetup {
         vm.stopPrank();
     }
 
+    function testAllowanceIssue() public {
```
we need to rename this test

dianacarvalho1 reviewed (approved):
amaziiingg
